### PR TITLE
Introduction of broker config cli and much more!

### PR DESCRIPTION
**Pull request by JacobCallahan**
The initial change of this was to introduce the broker config cli. However, to do this, a number of supporting and related changes needed to be made.

The first of these is the new ConfigManager class that manages reading from and writing to the config file.
Since so much logic was implemented for this, it made more sense to roll in the import/init functionality, that was originally in the settings module, into the new class.

This also gave me the opportunity to separate the test config from the example config. This removal, and other changes in this were supported by ConfigManager's migration functionality.

To simplify the nested chunk notation, I moved provider instances out from a list to just being a nested dictionary. This actually simplified the instance logic quite a bit.

Related to config simlification was the separation of the host settings to a new ssh chunk of the config. This is because not all users use the host functionality Broker provides. A likely change with this release is going to be removing a default ssh backend from Broker's requirements.

While I was at it, I also switched the CLI over to rich-click as part of a larger effort to improve the CLI experience.

### CODE REVIEW

File: broker/config_manager.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    def backup(self):
+        """Backup the current configuration file."""
+        logger.debug(
+            f"Backing up the configuration file to {self._settings_path.with_suffix('.bak')}"
+        )
+        self._settings_path.with_suffix(".bak").write_text(self._settings_path.read_text())
+
+    def restore(self):
+        """Restore the configuration file from a backup if it exists."""
+        logger.debug(
```
Comment (JacobCallahan) [inline broker/config_manager.py:120]: I'm debating whether this should be info level or not. There is a stronger case for this being info level than the message in backup.

File: broker/config_manager.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+    def get(self, chunk=None, curr_chunk=None):
+        """Get a chunk of Broker's config or the whole config."""
+        if not curr_chunk:
+            curr_chunk = self._cfg
+        if not chunk:
+            return curr_chunk
+        if C_SEP in chunk:
+            curr, chunk = chunk.split(C_SEP, 1)
+            # curr = int(curr) if curr.isdigit() else curr
```
Comment (JacobCallahan) [inline broker/config_manager.py:147]: This is a remnant from when I allowed traversing lists, like how instances are currently stored. Could keep it in for some future use or just get rid of these lines.

File: broker/config_manager.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            raw_data = self._import_config(GH_CFG, is_url=True)
+        if not raw_data:
+            raise exceptions.ConfigurationError(
+                f"Broker settings file not found at {self._settings_path.absolute()}."
+            )
+        chunk_data = self.get(chunk, yaml.safe_load(raw_data))
+        edited_chunk = self._interactive_edit(chunk_data)
+        self.update(chunk, edited_chunk)
+
+    def migrate(self):
```
Comment (JacobCallahan) [inline broker/config_manager.py:215]: i've considered whether we should allow people to force migrations for a target version.

File: broker/hosts.py [hunk 4]
```diff
@@ -45,6 +46,11 @@ def __init__(self, **kwargs):
             ipv6 (bool): Whether or not to use IPv6. Defaults to False.
             ipv4_fallback (bool): Whether or not to fallback to IPv4 if IPv6 fails. Defaults to True.
         """
+        global SETTINGS_VALIDATED  # noqa: PLW0603
+        if not SETTINGS_VALIDATED:
+            logger.debug("Validating ssh settings")
+            settings.validators.validate(only="SSH")
+            SETTINGS_VALIDATED = True
```
Comment (JacobCallahan) [inline broker/hosts.py:53]: delay validation until needed to avoid requiring the ssh config chunk.
this isn't currently important since every setting under ssh has a default value defined in its validator.

File: broker/hosts.py [hunk 5]
```diff
@@ -110,6 +117,8 @@ def connect(
             ipv6 (bool): Whether or not to use IPv6. Defaults to False.
             ipv4_fallback (bool): Whether or not to fallback to IPv4 if IPv6 fails. Defaults to True.
         """
+        from broker.session import Session
```
Comment (JacobCallahan) [inline broker/hosts.py:120]: delay import so we can potentially remove the requirement to have a ssh backend installed.

File: broker/settings.py [hunk 6]
```diff
@@ -97,23 +37,36 @@ def init_settings_from_local_repo(settings_path, interactive=False):
...
+
+if cfg_manager._get_migrations() and not TEST_MODE:
+    if INTERACTIVE_MODE:
+        click.secho(
+            "Broker settings file has pending migrations.\n"
+            "Continuing without running the migrations may cause errors.",
+            fg="red",
+        )
+        if click.confirm("Would you like to run the migrations now?"):
+            cfg_manager.migrate()
```
Comment (JacobCallahan) [inline broker/settings.py:54]: right now this is a choice, but we could also just remove the choice and make it automatic.

**Comment by JacobCallahan** [comment]
@tpapaioa @ogajduse I'd like to have this merged this week; any thoughts?

File: broker/config_manager.py (hunk 3 above)
Comment (JacobCallahan) [reply broker/config_manager.py]: we can now force

**Review by ogajduse — COMMENTED** [review]
This is a partial review, I went through a quarter of the patch and left some comments.

**Review by tpapaioa — APPROVED** [review]

File: broker/commands.py [hunk 7]
```diff
@@ -167,33 +178,38 @@ def provider_cmd(ctx, *args, **kwargs):  # the actual subcommand
...
+        table.add_column("Location", justify="left", style="magenta")
+
+        table.add_row("Broker Directory", str(settings.BROKER_DIRECTORY.absolute()))
+        table.add_row("Settings File", str(settings.settings_path.absolute()))
+        table.add_row("Inventory File", f"{settings.BROKER_DIRECTORY.absolute()}/inventory.yaml")
+        table.add_row("Log File", f"{settings.BROKER_DIRECTORY.absolute()}/logs/broker.log")
+
+        # Print the table
+        console = Console()
+        console.print(table)
```
[the remark this answers is not in the document]
Comment (JacobCallahan) [reply broker/commands.py:212]: good catch!

File: broker/config_manager.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        if self._interactive_mode is None:
+            self._interactive_mode = False
+            # GitHub action context
+            if "GITHUB_WORKFLOW" not in os.environ:
+                # determine if we're being ran from a CLI
+                for frame in inspect.stack()[::-1]:
+                    if "/bin/broker" in frame.filename:
+                        self._interactive_mode = True
+                        break
+        return self._interactive_mode
```
[the remark this answers is not in the document]
Comment (JacobCallahan) [reply broker/config_manager.py:74]: `sys.flags.interactive` is more for when python is ran with `-i`, so I don't want to go with that. However, I never thought about using `sys.stdin.isatty()` and that's a great suggestion. So I'm going to replace the post-github check with that.

File: broker/config_manager.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+    def _import_config(self, source, is_url=False):
+        """Initialize the broker settings file from a source."""
+        proceed = True
+        if self.interactive_mode:
+            try:
+                proceed = click.confirm(f"Get example file from {source}?")
+            except click.core.Abort:
+                # We're likely in a different non-interactive environment (container?)
+                self._interactive_mode = False
```
[the remark this answers is not in the document]
Comment (JacobCallahan) [reply broker/config_manager.py:101]: it will likely never get hit, but just in case....

File: broker/providers/ansible_tower.py [hunk 10]
```diff
@@ -21,6 +22,8 @@
 from broker import helpers
 from broker.providers import Provider
 
+YAML().register_class(awxkit.utils.PseudoNamespace)
```
Comment (JacobCallahan) [inline broker/providers/ansible_tower.py:25]: Actually, it's probably better to act on `helpers.yaml` here
```suggestion
helpers.yaml.register_class(awxkit.utils.PseudoNamespace)
```

File: broker/providers/ansible_tower.py [hunk 11]
```diff
@@ -778,9 +781,9 @@ def release(self, name, broker_args=None):
...
-def awxkit_representer(dumper, data):
-    """In order to resolve awxkit objects, a custom representer is needed."""
-    return dumper.represent_dict(dict(data))
+# def awxkit_representer(dumper, data):
+#     """In order to resolve awxkit objects, a custom representer is needed."""
+#     return dumper.represent_dict(dict(data))
 
 
-yaml.add_representer(awxkit.utils.PseudoNamespace, awxkit_representer)
+# yaml.add_representer(awxkit.utils.PseudoNamespace, awxkit_representer)
```
Comment (ogajduse) [inline broker/providers/ansible_tower.py:789]: Can this chunk be deleted? We can restore it anytime from a specific git ref.

File: broker/settings.py [hunk 12]
```diff
@@ -6,88 +6,28 @@
...
-    # determine if we're being ran from a CLI
-    for frame in inspect.stack()[::-1]:
-        if "/bin/broker" in frame.filename:
-            INTERACTIVE_MODE = True
-            break
-
-
+INTERACTIVE_MODE = ConfigManager().interactive_mode
 BROKER_DIRECTORY = Path.home().joinpath(".broker")
+TEST_MODE = "pytest" in sys.modules
```
Comment (ogajduse) [inline broker/settings.py:26]: Did you test that `pytest` is not in `sys.modules` when broker is imported in robottelo or a similar project?

File: tests/test_config_manager.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
+"""Test file for broker.config_manager module."""
+from ruamel.yaml import YAML
+from broker.config_manager import ConfigManager, GH_CFG
+from broker.settings import settings_path
+
+
+yaml = YAML()
```
Comment (ogajduse) [inline tests/test_config_manager.py:7]: As you suggested in another module, this could be changed to `from broker.helpers import yaml`.

File: broker/config_manager.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    @property
+    def interactive_mode(self):
+        """Determine if Broker is running in interactive mode."""
+        if self._interactive_mode is None:
+            self._interactive_mode = False
+            # GitHub action context
+            if "GITHUB_WORKFLOW" not in os.environ:
+                # determine if we're being ran from a CLI
+                if sys.stdin.isatty():
+                    self._interactive_mode = True
```
Comment (ogajduse) [inline broker/config_manager.py:67]: Doesn't `sys.stdin.isatty()` return False if it is being run from GH Actions?
```suggestion
            # determine if we're being run from a CLI
            if sys.stdin.isatty():
                self._interactive_mode = True
```

File: broker/config_manager.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ogajduse) [reply broker/config_manager.py]: I have created a new thread on the same topic. Let's discuss it there.

File: broker/config_manager.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+        logger.info("Config migration complete.")
+
+    def validate(self, chunk, providers=None):
+        """Validate a top-level chunk of Broker's config."""
+        if chunk == "all":
+            all_settings = [prov for prov in providers if prov != "TestProvider"] + ["base", "ssh"]
+            for item in all_settings:
+                self.validate(item, providers)
+            return
+        chunk = chunk.split(".")[0] if "." in chunk else chunk
```
[the remark this answers is not in the document]
Comment (JacobCallahan) [reply broker/config_manager.py:259]: good catch!

File: broker/providers/ansible_tower.py (hunk 11 above)
Comment (JacobCallahan) [reply broker/providers/ansible_tower.py:789]: yep, definitely should have been

File: broker/settings.py (hunk 12 above)
Comment (JacobCallahan) [reply broker/settings.py:26]: it almost certainly should trigger test mode, so we wouldn't automatically migrate while in a pytest session.

File: tests/test_config_manager.py (hunk 13 above)
Comment (JacobCallahan) [reply tests/test_config_manager.py:7]: for this, we don't need to modify the shared yaml object, so using it directly here to load is fine.

File: broker/config_manager.py (hunk 14 above)
Comment (JacobCallahan) [reply broker/config_manager.py:67]: unsure, but could test it by adding this to the workflow temporarily
```
python -c "import sys; print(f'{sys.stdin.isatty()=}')"
```

Comment (JacobCallahan) [reply broker/config_manager.py:67]: update: it is indeed false, based on that test.
```
sys.stdin.isatty()=False
```

Comment (JacobCallahan) [reply broker/config_manager.py]: and now completely removed that property in favor of a class attribute, thanks!

File: broker/settings.py (hunk 12 above)
Comment (ogajduse) [reply broker/settings.py:27]: Let me rephrase, do we want to have `TEST_MODE` set to `True` when we run robottelo (Foreman) tests? If we don't, this logic needs to be changed.
`pytest --setup-plan tests/foreman/api/test_ping.py -s`
```python
[74] > /home/ogajduse/repos/broker/broker/settings.py(28)<module>()
-> TEST_MODE = "pytest" in sys.modules
   2 frames hidden (try 'help hidden_frames')
(Pdb++) l
 23  	from broker.exceptions import ConfigurationError
 24  	
 25  	INTERACTIVE_MODE = ConfigManager.interactive_mode
 26  	BROKER_DIRECTORY = Path.home().joinpath(".broker")
 27  	breakpoint()
 28  ->	TEST_MODE = "pytest" in sys.modules
 29  	
 30  	if TEST_MODE:  # when in test mode, don't use the real broker directory
 31  	    BROKER_DIRECTORY = Path("tests/data/")
 32  	elif "BROKER_DIRECTORY" in os.environ:
 33  	    envar_location = Path(os.environ["BROKER_DIRECTORY"])
(Pdb++) "pytest" in sys.modules
True
```

File: broker/session.py [hunk 16]
```diff
@@ -19,7 +19,7 @@
 from broker.settings import settings
 
 SSH_BACKENDS = ("ssh2-python", "ssh2-python312", "ansible-pylibssh", "hussh")
-SSH_BACKEND = settings.SSH_BACKEND
+SSH_BACKEND = settings.BACKEND
```
Comment (tpapaioa) [inline broker/session.py:22]:
```suggestion
SSH_BACKEND = settings.SSH.BACKEND
```

**Review by tpapaioa — CHANGES_REQUESTED** [review]

File: broker/session.py (hunk 16 above)
Comment (JacobCallahan) [reply broker/session.py]: fixed

**Review by ogajduse — APPROVED** [review]
